Admin site

LucRome opened the pull request:
closes #27
closes #49
closes #50

codingHahn commented on eCourse_backend/courses/urls.py:
```diff
@@ -19,9 +19,22 @@
-    path('detail/<id>', view_course, name='view_course'),
+    # course
+    path(
+        'course/overview/page<int:page>',
```
Please remove the leading `course`.

codingHahn commented on eCourse_backend/courses/urls.py:
```diff
@@ -19,9 +19,22 @@
+        course_overview,
+        name='course_overview'),
+    # create user iframes
+    path('course/iframes/studentlist/<id>', course_student_list_iframe,
```
Please remove the leading `course`.

codingHahn commented on eCourse_backend/courses/urls.py:
```diff
@@ -19,9 +19,22 @@
+
+    path('edit/<id>', edit_course, name='edit_course'),
+
+    path('course/view/<id>', view_course, name='view_course')
```
Please remove the leading `course`.

codingHahn commented on eCourse_backend/courses/urls.py:
```diff
@@ -19,9 +19,22 @@
+    # admin: create course
+    path('admin/create/', create_course_admin, name="create_course_admin"),
+
+    path('edit/<id>', edit_course, name='edit_course'),
```
`edit` already exists in `alter`

LucRome replied:
should be resolved

codingHahn commented on eCourse_backend/courses/urls.py:
```diff
@@ -19,9 +19,22 @@
+    path('course/iframes/studentlist/<id>', course_student_list_iframe,
+         name='course_studentlist_iframe'),
+    # admin: create course
+    path('admin/create/', create_course_admin, name="create_course_admin"),
```
Why do we need an additional admin course creation?

codingHahn commented on eCourse_backend/courses/views.py:
```diff
@@ -1,25 +1,82 @@
         if request.user.type == 3:
             courses = Course.objects.all()
         else:
+            # courses = Course.objects.filter(student = user_id)
```
Please remove the commented out code

codingHahn commented on eCourse_backend/courses/views.py:
```diff
@@ -1,25 +1,82 @@
+    if request.method == 'GET':
+        course = get_object_or_404(Course, pk=id)
+        lecturer = course.lecturer_id
+        # print('lecturer', lecturer)
```
Please remove the commented out code

codingHahn commented on eCourse_backend/courses/views.py:
```diff
@@ -60,6 +117,26 @@ def create_course(request):
+    }
+    return render(request, 'courses/create_course_admin.html', context)
+
+
```
This seems very similar to `create_course`. What functionality do you miss from the existing function?

codingHahn commented on eCourse_backend/courses/views.py:
```diff
@@ -82,3 +159,17 @@ def delete_course(request, id):
+        course_object = get_object_or_404(Course, pk=id)
+        form = CourseForm(model_to_dict(course_object))
+
+    return render(request, 'admin/edit_course.html', {'form': form})
```
`alter_course` already does the same thing

codingHahn commented on eCourse_backend/eCourse_backend/settings.py:
```diff
@@ -43,11 +43,11 @@
     'django.contrib.staticfiles',
     'django_extensions',
-    'widget_tweaks',
     'eCourse_backend',
     'users',
     'file_exchange',
     'courses',
+    'widget_tweaks'
```
Is this necessary? If not, please revert this change. It pollutes the history

codingHahn commented on eCourse_backend/eCourse_backend/static/Css/basic_definitions.css:
```diff
@@ -0,0 +1,31 @@
+.input {
```
Please use lower case names for directories

codingHahn commented on eCourse_backend/eCourse_backend/static/readme.md:
```diff
@@ -0,0 +1,12 @@
+# Readme
+
+* **/Css:**
+    * Folder contains css files that are served as static files
+* **./:**
+    * Contains:
+        * different Versions of the Logo + some other files
+            * used to create compability with different browser features (some are just for Android/Windows/...)
+            * other files are also needed for these features (e.g. site.webmanifest)
+            * created with: [RealFaviconGenerator](https://realfavicongenerator.net/)
+        * the original logo
+            * created with: [FreeLogoDesign](https://de.freelogodesign.org)
```
This file is publically accessible. Is that wanted?

LucRome replied:
At the moment it doesn't hurt and otherwise it would be quite confusing why these files are there
I would leave it there since djangos static file hosting shouldn't be used in a production environment anyways

codingHahn commented on eCourse_backend/templates/admin/edit_course.html:
```diff
@@ -0,0 +1,22 @@
+  {% comment %} Button: Änderungen Speichern {% endcomment %}
+  <button class="btn btn-danger" type="submit">Speichern</button>
+</form>
+{% endblock %}
```
We already have `alter_courses`

codingHahn commented on eCourse_backend/templates/admin/home_Admin.html:
```diff
@@ -0,0 +1,5 @@
+{% extends 'home/home_auth.html' %}
```
Please use lower case file names

codingHahn commented on eCourse_backend/templates/admin/users/iframes/create_user/create_template.html:
```diff
@@ -0,0 +1,34 @@
+{% extends 'iframe_template.html' %}
```
Could you rename this file? To me at least, it isn't clear what `create_template` is.

codingHahn commented on eCourse_backend/templates/base.html:
```diff
@@ -6,11 +6,20 @@
 
 <head>
-  <meta charset="utf-8">
+    <meta charset="utf-8">
```
Please fix indentation

codingHahn commented on eCourse_backend/templates/home/home_notAuth.html:
```diff
@@ -0,0 +1,12 @@
+{% extends 'base.html' %}
```
Please use only lower case file names

codingHahn commented on eCourse_backend/users/urls.py:
```diff
@@ -19,8 +19,34 @@
+    # admin + staff list (iframe)
+    path('admin_s/iframes/adminstafflist/page<int:page>',
+         staff_admin_list_iframe, name='adminlist_admin_iframe'),
 ]
```
Please revise the site structure. We don't need extra `admin_s` functions. The normal `create` etc functions can only be accessed by an office user (admin) anyways. What is the greater goal here? But I understand the value of having extra pages for all user types.

We also need to talk about the `iframe/` url path naming. I think that is a good idea, but if we implement this, we should make this change in the whole codebase at once.

codingHahn commented on eCourse_backend/users/views.py:
```diff
@@ -60,3 +65,256 @@ def alter_user(request, id):
+
+# IFrames
+# create lecturer
+
```
Isn't that exactly the same as the `users/overview` page?

codingHahn commented on eCourse_backend/users/views.py:
```diff
@@ -60,3 +65,256 @@ def alter_user(request, id):
+    return render(
+        request,
+        'admin/users/iframes/deleted_user_iframe.html',
+        context)
```
This functionality already exists in `delete_user`

LucRome replied:
should be resolved

codingHahn commented on eCourse_backend/users/views.py:
```diff
@@ -60,3 +65,256 @@ def alter_user(request, id):
+    return render(
+        request,
+        'admin/users/iframes/edit_user_modalcontent.html',
+        context)
```
Please state which use cases the already existing view `alter_users` did not fullfill. Please extend the existing function

codingHahn commented on eCourse_backend/users/views.py:
```diff
@@ -60,3 +65,256 @@ def alter_user(request, id):
+    return render(
+        request,
+        'admin/users/iframes/staff_admin_list.html',
+        context)
```
The three views are very similar. Could this be merged into one view?

codingHahn commented on eCourse_backend/users/views.py:
```diff
@@ -60,3 +65,256 @@ def alter_user(request, id):
+    return render(
+        request,
+        'admin/users/iframes/create_user/create_officeuser.html',
+        context)
```
The three views are very similar. Could this be merged into one view?
Furthermore, the `create_user` function has much of the same functionality in fewer lines of code

codingHahn commented on eCourse_backend/users/forms.py:
```diff
@@ -1,8 +1,36 @@
+class AdminStaffFilterForm(Form):
+    username = CharField(max_length=50, label="Username", required=False)
+    first_name = CharField(max_length=50, label="Vorname", required=False)
+    last_name = CharField(max_length=50, label="Nachname", required=False)
```
Those forms are very similar. Could they be merged?

codingHahn commented on eCourse_backend/users/forms.py:
```diff
@@ -1,8 +1,36 @@
+        model = User
+        fields = ['username', 'first_name', 'last_name',
+                  'email', 'matr_nr', 'is_superuser']
+
```
Please don't declare regular users as superusers, because they can put the database in a bad state. They are meant for maintenance tasks.

codingHahn reviewed (changes requested):
Please address the points mentioned in the review. Please remember that filenames and directories should be snake case.

phie73 commented on eCourse_backend/users/forms.py:
```diff
@@ -1,8 +1,27 @@
+    class Meta:
+        model = User
+        fields = ['username', 'first_name', 'last_name',
+                  'email', 'matr_nr']
```
these two forms are exactly the same

phie73 commented on eCourse_backend/users/urls.py:
```diff
@@ -19,8 +19,29 @@
+         lecturer_list_iframe, name='lecturerlist_admin_iframe'),
+    # admin + staff list (iframe)
+    path('admin_s/iframes/adminstafflist/page<int:page>',
+         staff_admin_list_iframe, name='adminlist_admin_iframe'),
```
please remove admin_s

phie73 commented on eCourse_backend/users/views.py:
```diff
@@ -1,62 +1,258 @@
+
+
+# Views for the admin
+# TODO: check whethter user is from the right group
```
please remove the todo because the permissions already solve this point

phie73 commented on eCourse_backend/users/views.py:
```diff
@@ -1,62 +1,258 @@
     if request.method == 'GET':
-        all_users = User.objects.all()
-        return render(request, 'users/overview.html', {'users': all_users})
+        return render(request, 'admin/users/administration.html', {})
```
by the function name their purpose isn't clear please rename

codingHahn commented:
When I log in this happens: 
![image](https://user-images.githubusercontent.com/16626332/116516230-a67e7700-a8cd-11eb-8892-2ab870f3bd29.png)

phie73 commented on eCourse_backend/courses/urls.py:
```diff
@@ -19,9 +19,14 @@
+    path('iframes/studentlist/<id>', course_student_list_iframe,
+         name='course_studentlist_iframe'),
+    # admin: create course
+    path('admin/create/', create_course_admin, name="create_course_admin"),
```
please remove the admin from 'admin/creare/'

phie73 commented on eCourse_backend/courses/views.py:
```diff
@@ -1,16 +1,51 @@
+        'students': students,
+    }
+    return render(request, 'courses/iframes/course_student_list.html', context)
+
```
this function already exists on the master branch see courses/views.py view_course on the master branch. The functionality is extended for mapping the files and exercises to the courses. This function used to exist in this file (see below) but it was deleted. If you need additional data please add this to the existing function or create an issue.

phie73 commented on eCourse_backend/courses/views.py:
```diff
@@ -19,50 +54,50 @@ def course_overview(request):
 @login_required
 @permission_required('courses.create_course', raise_exception=True)
-def create_course(request):
+def create_course_admin(request):
```
please revert the name change or explain why it is needed.

phie73 commented on eCourse_backend/users/views.py:
```diff
@@ -1,62 +1,258 @@
+@xframe_options_exempt
+@login_required
+@permission_required('users.manage_users', raise_exception=True)
+def delete_user_iframe(request, username):
```
Please keep the user id instead of the username for consistency reasons

phie73 commented on eCourse_backend/users/views.py:
```diff
@@ -1,62 +1,258 @@
 @login_required
 @permission_required('users.manage_users', raise_exception=True)
-def alter_user(request, id):
+def edit_user_admin_modalcontent_iframe(request, username):
```
Please keep the user id instead of the username for consistency reasons.

phie73 commented on eCourse_backend/courses/views.py:
```diff
@@ -1,16 +1,51 @@
+        filter_form = CourseStudentFilterForm()
+        course = get_object_or_404(Course, pk=id)
+        lecturer = course.lecturer
+        # print('lecturer', lecturer)
```
could you please remove this debug info?

phie73 commented on eCourse_backend/users/views.py:
```diff
@@ -1,62 +1,257 @@
         user_form = UserForm(request.POST)
         if user_form.is_valid():
-            user_form.save()
+            Lecturer.objects.create_user(
+                username=user_form['username'].data,
+                email=user_form['email'].data,
+                first_name=user_form['first_name'].data,
+                last_name=user_form['last_name'].data,
```
please use validated data, this is vulnerable looking at sql injections

LucRome replied:
How could that be achieved??
Could you please demonstrate the changes as I don't know exactly how it should be done

phie73 replied:
we posted an example in the discord channel, sorry for not mentioning here

phie73 commented on eCourse_backend/users/views.py:
```diff
@@ -1,62 +1,257 @@
+                username=user_form['username'].data,
+                email=user_form['email'].data,
+                first_name=user_form['first_name'].data,
+                last_name=user_form['last_name'].data,
```
please use validated data, this is vulnerable looking at sql injections

phie73 commented on eCourse_backend/users/views.py:
```diff
@@ -1,62 +1,257 @@
+                first_name=user_form['first_name'].data,
+                last_name=user_form['last_name'].data,
+                matr_nr=user_form['matr_nr'].data
+            )
```
please use validated data, this is vulnerable looking at sql injections

phie73 commented on eCourse_backend/users/views.py:
```diff
@@ -1,62 +1,257 @@
+                matr_nr__contains=filter_form['matr_nr'].data,
+                first_name__contains=filter_form['first_name'].data,
+                last_name__contains=filter_form['last_name'].data,
+                username__contains=filter_form['username'].data)
```
please use validated data, this is vulnerable looking at sql injections

phie73 commented on eCourse_backend/users/views.py:
```diff
@@ -1,62 +1,257 @@
+            lecturers = Lecturer.objects.filter(
+                first_name__contains=filter_form['first_name'].data,
+                last_name__contains=filter_form['last_name'].data,
+                username__contains=filter_form['username'].data)
```
please use validated data, this is vulnerable looking at sql injections

phie73 commented on eCourse_backend/users/views.py:
```diff
@@ -1,62 +1,257 @@
+            request.POST, )
+        if filter_form.is_valid():
+            # Filter
+            users = Office.objects.filter(
+                first_name__contains=filter_form['first_name'].data,
+                last_name__contains=filter_form['last_name'].data,
+                username__contains=filter_form['username'].data)
```
please use validated data, this is vulnerable looking at sql injections

LucRome replied:
Also please at least provide an Example, I don't know how else it could be done
And also what do you mean by validated data? Since we call the .is_valid() function.

43-50ph14 reviewed (changes requested):
looks good overall, please address the minor issues we found. Looking forward to merge the pr when this small changes are done.

43-50ph14 reviewed (approved):
lets solve the merge conflicts together on Monday?

LucRome commented:
merge conflicts should be solved:

- the Logo.png should be the newest Version
- in `courses/views.py` there was
    - a conflict regarding the imports -> all imports were kept
    - a conflict regarding the two view function for the course overview
        -> both functions kept (also the one from the backend, since it introduced new functions, which should be used by the frontend, should be merged and deleted later on)
- the template `courses/detail.html` was kept, despite not being used by frontend, since it also introduced some new functions
